src/java/org/apache/cassandra/service/accord/serializers/TopologySerializers.java

```diff
@@ -101,96 +77,85 @@ public <V> Node.Id deserialize(ByteBuffer src, int position)
+            CollectionSerializers.serializeList(shard.joining, out, nodeId);
             out.writeUnsignedVInt32(shard.flags().bitset());
         }
 
         @Override
-        public Shard deserialize(DataInputPlus in, int version) throws IOException
+        public Shard deserialize(DataInputPlus in) throws IOException
         {
-            Range range = ShardSerializer.this.range.deserialize(in, version);
-            SortedArrayList<Node.Id> nodes = CollectionSerializers.deserializeSortedArrayList(in, version, nodeId, Node.Id[]::new);
-            SortedArrayList<Node.Id> notInFastPath = CollectionSerializers.deserializeSortedArrayList(in, version, nodeId, Node.Id[]::new);
-            SortedArrayList<Node.Id> joining = CollectionSerializers.deserializeSortedArrayList(in, version, nodeId, Node.Id[]::new);
+            Range range = ShardSerializer.this.range.deserialize(in);
+            SortedArrayList<Node.Id> nodes = CollectionSerializers.deserializeSortedArrayList(in, nodeId, Node.Id[]::new);
+            SortedArrayList<Node.Id> notInFastPath = CollectionSerializers.deserializeSortedArrayList(in, nodeId, Node.Id[]::new);
+            SortedArrayList<Node.Id> joining = CollectionSerializers.deserializeSortedArrayList(in, nodeId, Node.Id[]::new);
             int flags = in.readUnsignedVInt32();
             return Shard.SerializerSupport.create(range, nodes, notInFastPath, joining, new TinyEnumSet<>(flags));
         }
 
         @Override
-        public long serializedSize(Shard shard, int version)
+        public long serializedSize(Shard shard)
         {
-            long size = range.serializedSize(shard.range, version);
-            size += CollectionSerializers.serializedListSize(shard.nodes, version, nodeId);
-            size += CollectionSerializers.serializedListSize(shard.notInFastPath, version, nodeId);
-            size += CollectionSerializers.serializedListSize(shard.joining, version, nodeId);
+            long size = range.serializedSize(shard.range);
+            size += CollectionSerializers.serializedListSize(shard.nodes, nodeId);
+            size += CollectionSerializers.serializedListSize(shard.notInFastPath, nodeId);
+            size += CollectionSerializers.serializedListSize(shard.joining, nodeId);
             size += TypeSizes.sizeofUnsignedVInt(shard.flags().bitset());
             return size;
         }
-    };
+    }
 
-    public static final IVersionedSerializer<Topology> topology = new IVersionedSerializer<>()
+    public static final Serializer<Topology> topology = new Serializer<>()
     {
         @Override
-        public void serialize(Topology topology, DataOutputPlus out, int version) throws IOException
+        public void serialize(Topology topology, DataOutputPlus out) throws IOException
         {
             out.writeLong(topology.epoch());
-            ArraySerializers.serializeArray(topology.unsafeGetShards(), out, version, shard);
```

Comment:
`unsafeGetShards` is unsafe and causes this to fail.  we have a `local` and `global` topology we serialize, but the `local` topology comes back as `global` when we deserialize!  moving to `shards` makes sure we respect this.